Documentation of new Dart SDK features (#145)
configcat/docs: website/docs/sdk-reference/dart.md

| `mode`                      | Optional, sets the polling mode for the client. [More about polling modes](#polling-modes). |
| `logger`                    | Optional, sets the internal logger and log level. [More about logging](#logging). |
| `override`                  | Optional, configures local feature flag & setting overrides. [More about feature flag overrides](#flag-overrides). |

> | `defaultUser`               | Optional, configures the default user. [More about default user](#default-user). |

Reviewer: please try to avoid the word `configure` and `configuration`

Author: done

);
```

## Anatomy of `getValueDetails()`

> `getValueDetails()` is similar to `getValue()` but instead of returning only the evaluated value it gives more detailed information about the evaluation result.

Author: ✅

| Parameters     | Description                                                                                                  |
| -------------- | ------------------------------------------------------------------------------------------------------------ |
| `key`          | **REQUIRED.** The key of a specific setting or feature flag. Set on *ConfigCat Dashboard* for each setting.                           |
| `defaultValue` | **REQUIRED.** This value will be returned in case of an error.                                               |
| `user`         | Optional, *User Object*. Essential when using Targeting. [Read more about Targeting.](advanced/targeting.md) |
```dart
final details = await client.getValueDetails(
    key: 'keyOfMySetting', 
    defaultValue: false,
    user: ConfigCatUser(identifier: '#USER-IDENTIFIER#'), // Optional User Object
);

// Details holds the following information:
details.value;                              // The evaluated value of the feature flag or setting.
details.key;                                // The key of the evaluated feature flag or setting.

> details.isDefaultValue;                     // True when the default value passed to getValueDetails() is returned due to an error.

Reviewer 2: Is `isDefaultValue` `false` if the default value is returned but there was no error?

Author: Technically there's no such case, we always have an error if we decide to return with the default value.

Reviewer 2: I'm asking from the perspective of the user. If I'm the user and I have a few targeting rules to handle exceptional cases, and I have a default value for everything else on the dashboard. Then I feel tempted to check the isDefaultVale to see if the "default value" was returned. 

For example, if I have a flag like this:

canUseConfigCat:
- IF USER's Email IS IN "hacker@qq.com, 123456@qq.com" THEN `FALSE`
- ELSE IF USER's Country IS IN "China" THEN `FALSE`
- ELSE `TRUE`

Here, I think about `TRUE` as the default value.

Author: The `isDefaultValue` doesn't indicate that, it's just for signaling that the passed default value was returned. 
The indicator for the case you mentioned (no rules were used for evaluation) is that both `matchedEvaluationRule` and `matchedEvaluationRule` is null.
If I remember correctly, we agreed on not including this case as 'default value' as it's just an internal term used by us.

> details.error;                              // In case of an error this field will contain its message.

Author: ✅

> details.matchedEvaluationPercentageRule;    // When the evaluation was based on a percentage rule, this field will contain that specific rule.

Reviewer 2: If the evaluation was based on a percentage rule, this field contains that specific rule.

Author: ✅

> details.matchedEvaluationRule;              // When the evaluation was based on an targeting rule, this field will contain that specific rule.

Reviewer 2: If the evaluation was based on a targeting rule, this field contains that specific rule.

- if
- an => a
- I prefer removing "will"

Author: ✅

> details.fetchTime;                          // The last download time of the current config.

Reviewer 2: Time like what?
Second? Millisecond? DateTimeUtc?

);
```

## Anatomy of `getValueDetails()`

`getValueDetails()` is similar to `getValue()` but instead of returning only the evaluated value it gives more detailed information about the evaluation result.

> 

Reviewer: example here would be great to demonstrate what "more detailed" means

Author: It's there a few lines below in the usage sample

| Parameters     | Description                                                                                                  |
| -------------- | ------------------------------------------------------------------------------------------------------------ |

> | `key`          | **REQUIRED.** Setting-specific key. Set on *ConfigCat Dashboard* for each setting.                           |

Reviewer: ```suggestion
| `key`          | **REQUIRED.** The key of a specific setting or feature flag. Set on *ConfigCat Dashboard* for each setting.                           |
```

| `defaultValue` | **REQUIRED.** This value will be returned in case of an error.                                               |

> | `user`         | Optional, *User Object*. Essential when using Targeting. [Read more about Targeting.](advanced/targeting.md) |

Reviewer: ```suggestion
| `userObject`         | Optional, *User Object*. Essential when using Targeting. [Read more about Targeting.](advanced/targeting.md) |
```

why not call it as it's original name?

Author: because we call the parameter `user` in each SDK

```dart
final details = await client.getValueDetails(
    key: 'keyOfMySetting', 
    defaultValue: false,
    user: ConfigCatUser(identifier: '#USER-IDENTIFIER#'), // Optional User Object
);

// Details holds the following information:
details.value;                              // The evaluated value of the feature flag or setting.
details.key;                                // The key of the evaluated feature flag or setting.

> details.variationId;                        // The variation ID.

Reviewer: do we have a doc that we can link? or blogpost about what is the variation id

Reviewer: okay

);
```

### Default user

> There's an option to set a default user object that will be used at feature flag and setting evaluation. It can be useful when your application has usually a single user only, or rarely switches users.

Reviewer 2: I personally, would remove the word "usually".

Author: ✅

You can set the default user object either on SDK initialization:

```dart
final client = ConfigCatClient.get(
    sdkKey: '#YOUR-SDK-KEY#',
    options: ConfigCatOptions(
        defaultUser: ConfigCatUser(identifier: 'john@example.com')
    )
);
```
or with the `setDefaultUser()` method of the ConfigCat client.
```dart
client.setDefaultUser(ConfigCatUser(identifier: 'john@example.com'));
```

> If the default user is set, the SDK will use it every time when the `getValue()`, `getValueDetails()`, `getAllValues()`, or `getAllVariationIds()` methods are called without passing a user object to them.

Reviewer 2: A suggestion:

Whenever the `getValue()`, `getValueDetails()`, `getAllValues()`, or `getAllVariationIds()` methods are called without an explicit user object parameter, the SDK will automatically use the default user as a user object.

Author: ✅

);
```

### Default user

There's an option to set a default user object that will be used at feature flag and setting evaluation. It can be useful when your application has usually a single user only, or rarely switches users.

> You can set the default user either on SDK initialization:

Reviewer: ```suggestion
You can set the default user object either on SDK initialization:
```
This should be user object everywhere consequently

```dart
final client = ConfigCatClient.get(
    sdkKey: '#YOUR-SDK-KEY#',
    options: ConfigCatOptions(
        defaultUser: ConfigCatUser(identifier: 'john@example.com')
    )
);
```
or with the `setDefaultUser()` method of the ConfigCat client.
```dart
client.setDefaultUser(ConfigCatUser(identifier: 'john@example.com'));
```

> If the default user is set, the SDK will use it's value every time when the `getValue()`, `getValueDetails()`, `getAllValues()`, or `getAllVariationIds()` methods are called without an other user object.

Reviewer: ```suggestion
If the default user is set, the SDK will use it every time when the `getValue()`, `getValueDetails()`, `getAllValues()`, or `getAllVariationIds()` methods are called without passing a user object to them.
```

```
> `getValue()` returns `defaultValue` if the cache is empty. Call `forceRefresh()` to update the cache.

## Hooks

> With the following hooks you can subscribe particular events sent by the SDK:

Reviewer 2: - subscribe => subscribe to ("to" is missing)
- send => fired (a suggestion only)

Author: ✅

> - `onClientReady()`: This event is sent when the SDK reaches the ready state. When the SDK is configured with lazy load or manual polling it's considered ready right after instantiation.

Reviewer 2: ~~When~~ If the SDK is ...

Author: ✅

> When it's using auto polling, the ready state is reached when the SDK has a valid config.json loaded into memory either from cache or from HTTP. When the config couldn't be loaded neither from cache nor from HTTP the `onClientReady` event fires when the auto polling's `maxInitWaitTime` is reached.

Reviewer 2: - ~~When it's~~ If the SDK is using polling mode
- ~~When~~ If the config couldn't

Author: ✅

- `onConfigChanged(Map<string, Setting>)`: This event is sent when the SDK loads a valid config.json into memory from cache, and each subsequent time when the loaded config.json changes via HTTP.
- `onFlagEvaluated(EvaluationDetails)`: This event is sent each time when the SDK evaluates a feature flag or setting. The event sends the same evaluation details that you would get from [`getValueDetails()`](#anatomy-of-getvaluedetails).
- `error(String)`: This event is sent when an error occurs within the ConfigCat SDK.

You can subscribe to these events either on SDK initialization: 
```dart
final client = ConfigCatClient.get(

>     sdkKey: '<PLACE-YOUR-SDK-KEY-HERE>',

Reviewer 2: '#YOUR-SDK-KEY#' would be the consistent choice

Author: ✅

    options: ConfigCatOptions(
        mode: PollingMode.manualPoll(),
        hooks: Hooks(
            onFlagEvaluated: (details) => /* handle the event */
        )
    )
);
```
or with the `hooks` property of the ConfigCat client:
```dart
client.hooks.addOnFlagEvaluated((details) => /* handle the event */);
```

## Online / Offline mode

In cases when you'd want to prevent the SDK from making HTTP calls, you can put it in offline mode:
```dart
client.setOffline();
```

> In offline mode the SDK won't initiate HTTP requests and will work only from its cache.

Reviewer 2: comma is always need after the preposition:

In offline mode, the

```
> `getValue()` returns `defaultValue` if the cache is empty. Call `forceRefresh()` to update the cache.

## Hooks

With the following hooks you can subscribe particular events sent by the SDK:

- `onClientReady()`: This event is sent when the SDK reaches the ready state. When the SDK is configured with lazy load or manual polling it's considered ready right after instantiation.

> When it's using auto polling, the ready state is reached when the SDK has a valid configuration loaded into memory either from cache or from HTTP. When the config couldn't be loaded neither from cache nor from HTTP the `onClientReady` event fires when the auto polling's `maxInitWaitTime` is reached.

Reviewer: ```suggestion
When it's using auto polling, the ready state is reached when the SDK has a valid config.json (or something else) loaded into memory either from cache or from HTTP. When the config couldn't be loaded neither from cache nor from HTTP the `onClientReady` event fires when the auto polling's `maxInitWaitTime` is reached.
```

Author: wdym by `or something else`?

Reviewer: anything but configuration

Author: replaced configuration with config.json

> - `onConfigChanged(Map<string, Setting>)`: This event is sent when the SDK loads a valid configuration into memory from cache, and each subsequent time when the loaded configuration changes via HTTP.

Reviewer: `Configuration` should be replaced almost everywhere

Author: replaced configuration with config.json

- `onFlagEvaluated(EvaluationDetails)`: This event is sent each time when the SDK evaluates a feature flag or setting. The event sends the same evaluation details that you would get from [`getValueDetails()`](#anatomy-of-getvaluedetails).

> - `error(String)`: This event is sent when an error occurs in the SDK's functioning.

Reviewer: ```suggestion
- `error(String)`: This event is sent when an error occurs within the ConfigCat SDK.
```
